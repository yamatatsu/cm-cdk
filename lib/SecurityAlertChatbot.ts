import {
  App,
  Stack,
  StackProps,
  aws_chatbot,
  aws_sns,
  aws_events_targets,
  aws_events,
} from "aws-cdk-lib";

type Props = StackProps & {
  slackWorkspaceId: string;
  slackChannelId: string;
};

export class SecurityAlertChatbot extends Stack {
  constructor(scope: App, id: string, props: Props) {
    super(scope, id, props);

    const topic = new aws_sns.Topic(this, "GuardDutyTopic");

    const rule = new aws_events.Rule(this, "GuardDutyNotificationRule", {
      ruleName: "GuardDutyNotificationRule",
      description: "Alert to SNS topic when find threats by Guardduty",
      eventPattern: {
        source: ["aws.guardduty"],
        detailType: ["GuardDuty Finding"],
      },
    });
    rule.addTarget(new aws_events_targets.SnsTopic(topic));

    new aws_chatbot.SlackChannelConfiguration(
      this,
      "SlackChannelConfiguration",
      {
        slackChannelConfigurationName: "SecurityAlert",
        slackWorkspaceId: props.slackWorkspaceId,
        slackChannelId: props.slackChannelId,
        notificationTopics: [topic],
      }
    );
  }
}
