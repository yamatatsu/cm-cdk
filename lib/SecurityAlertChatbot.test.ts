import { SynthUtils } from "@aws-cdk/assert";
import { App, Stack, aws_sns as sns } from "aws-cdk-lib";
import { SecurityAlertChatbot } from "./SecurityAlertChatbot";

test("snapshot test", () => {
  const app = new App();
  const stack = new Stack(app, "testStack");

  const topic = new sns.Topic(stack, "test-Topic");

  const target = new SecurityAlertChatbot(app, "Target", {
    slackWorkspaceId: "test-slackWorkspaceId",
    slackChannelId: "test-slackChannelId",
  });

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot();
});
