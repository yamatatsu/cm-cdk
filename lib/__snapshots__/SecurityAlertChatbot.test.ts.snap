// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Resources": Object {
    "GuardDutyNotificationRule6F3FBB81": Object {
      "Properties": Object {
        "Description": "Alert to SNS topic when find threats by Guardduty",
        "EventPattern": Object {
          "detail-type": Array [
            "GuardDuty Finding",
          ],
          "source": Array [
            "aws.guardduty",
          ],
        },
        "Name": "GuardDutyNotificationRule",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Ref": "GuardDutyTopic003A9A5B",
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "GuardDutyTopic003A9A5B": Object {
      "Type": "AWS::SNS::Topic",
    },
    "GuardDutyTopicPolicy0CA49850": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
              "Resource": Object {
                "Ref": "GuardDutyTopic003A9A5B",
              },
              "Sid": "0",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": Array [
          Object {
            "Ref": "GuardDutyTopic003A9A5B",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
    "SlackChannelConfigurationC73E0866": Object {
      "Properties": Object {
        "ConfigurationName": "SecurityAlert",
        "IamRoleArn": Object {
          "Fn::GetAtt": Array [
            "SlackChannelConfigurationConfigurationRole066766EC",
            "Arn",
          ],
        },
        "SlackChannelId": "test-slackChannelId",
        "SlackWorkspaceId": "test-slackWorkspaceId",
        "SnsTopicArns": Array [
          Object {
            "Ref": "GuardDutyTopic003A9A5B",
          },
        ],
      },
      "Type": "AWS::Chatbot::SlackChannelConfiguration",
    },
    "SlackChannelConfigurationConfigurationRole066766EC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "chatbot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
